apps/bixarena/auth-service/src/main/java/org/sagebionetworks/bixarena/auth/service/api/AuthApiDelegateImpl.java

```diff
@@ -221,46 +217,53 @@ public ResponseEntity<Callback200ResponseDto> callback(String code, String state
+      log.info("OIDC callback: All ID token claims: {}", idClaims.keySet());
+      log.debug("OIDC callback: Full ID token claims: {}", idClaims);
+
       String sub = (String) idClaims.get("sub");
       if (sub == null) {
         log.info("OIDC callback: missing sub claim in id token");
         return ResponseEntity.status(400)
           .contentType(MediaType.APPLICATION_JSON)
           .body(Callback200ResponseDto.builder().status("error:missing-sub").build());
       }
+
       String email = (String) idClaims.get("email");
       String givenName = (String) idClaims.getOrDefault("given_name", null);
       String familyName = (String) idClaims.getOrDefault("family_name", null);
-      String preferred = (String) idClaims.getOrDefault("preferred_username", sub);
-      // Upsert user
-      UserEntity existingOrNew = userRepository
-        .findByUsername(preferred)
-        .orElseGet(() ->
-          UserEntity.builder()
-            .username(preferred)
-            .email(email)
-            .firstName(givenName)
-            .lastName(familyName)
-            .build()
-        );
-      UserEntity persistedUser = existingOrNew.getId() == null
-        ? userRepository.save(existingOrNew)
-        : existingOrNew;
-      // Link external account
-      externalAccountRepository
-        .findByProviderAndExternalId(Provider.synapse, sub)
-        .orElseGet(() ->
-          externalAccountRepository.save(
-            ExternalAccountEntity.builder()
-              .user(persistedUser)
-              .provider(Provider.synapse)
-              .externalId(sub)
-              .externalUsername(preferred)
-              .externalEmail(email)
-              .build()
-          )
-        );
+      String preferredUsername = (String) idClaims.getOrDefault("preferred_username", sub);
+      String userName = (String) idClaims.getOrDefault("user_name", null);
+
+      // Use user_name claim if available (this is the actual Synapse username)
+      // Otherwise fall back to preferred_username
+      String synapseUsername = userName != null ? userName : preferredUsername;
+
+      log.info(
+        "OIDC callback: ID token claims - sub={}, user_name={}, preferred_username={}, email={}, given_name={}, family_name={}",
+        sub,
+        userName,
+        preferredUsername,
+        email,
+        givenName,
+        familyName
+      );
+
+      log.info("OIDC callback: Using username={} for user creation/update", synapseUsername);
+
+      // Use UserService to handle user creation/update and login tracking
+      UserEntity persistedUser = userService.handleUserLogin(
+        Provider.synapse,
+        sub,
+        synapseUsername,
+        email,
+        givenName,
+        familyName
+      );
+
       // Establish authenticated session principal
       session.setAttribute("AUTH_SUBJECT", persistedUser.getUsername());
       session.setAttribute("AUTH_PREFERRED_USERNAME", persistedUser.getUsername());
```

Comment:
How is the authenticated user's user ID made available to the API service? Is it passed via session here?